impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,198 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
+
+            // Do some checks to determine it from vendor specific data
+            Principal vendorCallerPrincipal = getVendorCallerPrincipal(principal, false);
+            if (vendorCallerPrincipal != null) {
+                return vendorCallerPrincipal;
             }
+        }
 
-            return null;
+        return null;
+    }
+
+    /**
+     * Get the underlying caller principal based on vendor specific (e.g.: class
+     * names, EJB unauthenticated name, etc)
+     *
+     * @param principal
+     * @return
+     */
+    private Principal getVendorCallerPrincipal(Principal principal, boolean isEjb) {
+        switch (principal.getClass().getName()) {
+            case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
+                return isEjb ? getAuthenticatedPrincipal(principal, "ANONYMOUS") : principal;
+            case "com.ibm.ws.security.authentication.principals.WSPrincipal": // Liberty
+                return isEjb ? getAuthenticatedPrincipal(principal, "UNAUTHENTICATED") : principal;
+            // JBoss EAP/WildFly convention 1 - single top level principal of the below type
+            case "org.jboss.security.SimplePrincipal":
+                return isEjb ? getAuthenticatedPrincipal(principal, "anonymous") : principal;
+            // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
+            case "org.jboss.security.SimpleGroup":
+                if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
+
+                    Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
+
+                    if (groupMembers.hasMoreElements()) {
+                        return isEjb ? getAuthenticatedPrincipal(groupMembers.nextElement(), "anonymous") : principal;
+                    }
+                }
+                break;
+            case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
+                try {
+                    Principal tomeePrincipal = (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
+                            .getMethod("getUserPrincipal")
+                            .invoke(
+                                    Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
+                                            .getMethod("getTomcatPrincipal")
+                                            .invoke(principal));
+
+                    return isEjb ? getAuthenticatedPrincipal(tomeePrincipal, "guest") : tomeePrincipal;
+                } catch (Exception e) {
+
+                }
+                break;
         }
-        
+
         return null;
     }
-     
+
+    private Principal getAuthenticatedPrincipal(Principal principal, String anonymousCallerName) {
+        return !anonymousCallerName.equals(principal.getName()) ? principal : null;
```

Comment:
If `isEjb` is passed as a parameter here, the logic could be:
````
if (isEjb && principal != null && anonymousCallerName.equals(principal.getName())) {
    return null;
}
return principal;
````
Not quite as succinct as the ternary operator, but probably a little easier to read with the three tests that are needed.